src/Aspire.Hosting/Dashboard/ResourcePublisher.cs

```diff
@@ -81,7 +90,29 @@ public async ValueTask IntegrateAsync(ResourceViewModel resource, ResourceChange
 
         foreach (var channel in _outgoingChannels)
         {
-            await channel.Writer.WriteAsync(new(changeType, resource), cancellationToken).ConfigureAwait(false);
+            await channel.Writer.WriteAsync(new(changeType, ToViewModel(resource)), cancellationToken).ConfigureAwait(false);
         }
     }
+
+    private static ResourceViewModel ToViewModel(ResourceSnapshot snapshot)
+    {
+        // NOTE this is a temporary method -- ultimately, snapshots will be converted to gRPC messages,
+        // then on the other side those gRPC messages will be turned into ResourceViewModel objects, or
+        // maybe just used to update existing ones.
+
+        return new()
+        {
+            Name = snapshot.Name,
+            ResourceType = snapshot.ResourceType,
+            DisplayName = snapshot.DisplayName,
+            Uid = snapshot.Uid,
+            State = snapshot.State,
+            CreationTimeStamp = snapshot.CreationTimeStamp,
+            Environment = snapshot.Environment.Select(e => new EnvironmentVariableViewModel { Name = e.Name, Value = e.Value, FromSpec = e.FromSpec }).ToImmutableArray(),
+            Endpoints = snapshot.Endpoints,
+            Services = snapshot.Services.Select(s => new ResourceServiceViewModel(s.Name, s.AllocatedAddress, s.AllocatedPort)).ToImmutableArray(),
+            ExpectedEndpointsCount = snapshot.ExpectedEndpointsCount,
+            CustomData = snapshot.Data.ToFrozenDictionary(pair => pair.Key, pair => pair.Value, StringComparers.ResourceDataKey)
+        };
+    }
```

Comment:
This is the point where we go from snapshots to view models.

In the next PR, gRPC will be inserted here, with types generated from the `.proto` being used between.

```mermaid
flowchart LR;
  CustomResource -->ResourceSnapshot --> gRPC --> ResourceViewModel
```

Reply:
This has been replaced with gRPC in #1476.